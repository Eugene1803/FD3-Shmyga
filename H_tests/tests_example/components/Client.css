.ClientActive
{
  background-color: green;
}

.ClientBlocked
{
  background-color: red;
}

