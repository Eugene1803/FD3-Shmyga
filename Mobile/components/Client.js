import React from 'react';
import PropTypes from 'prop-types';
import {delEditEvents} from './events';
import './Client.css';

class Client extends React.PureComponent {
/*
  static propTypes = {
    info:PropTypes.shape({
      id: PropTypes.number.isRequired,
      fio: PropTypes.string.isRequired,
      balance: PropTypes.number.isRequired,
    }),
  };
*/
  state = {
    info: this.props.info,
  };

  deleteMe = () => {
    delEditEvents.emit('deleteClient',this.state.info.id);
  } 

  editMe = () => {
    delEditEvents.emit('editClient',this.state.info)
  }
  
  componentWillReceiveProps = (newProps) => {
    this.setState({info:newProps.info});
  };

  render() {

    console.log("MobileClient id="+this.state.info.id+" render");
    
    return (
      <tr>
        <td>{this.state.info.surname}</td>
        <td>{this.state.info.name}</td>
        <td>{this.state.info.secName}</td>
        <td>{this.state.info.balance}</td>
        <td className={(this.state.info.balance > 0)?'ClientActive':'ClientBlocked'}>{(this.state.info.balance > 0)?'Активен':'Заблокирован'}</td>
        <td><button onClick={this.editMe}>Редактировть</button></td>
        <td><button onClick={this.deleteMe}>Удалить</button></td>
      </tr>
    );

  }

}

export default Client;
