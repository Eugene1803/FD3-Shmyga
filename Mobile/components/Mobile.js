import React from 'react';
import PropTypes from 'prop-types';
import {delEditEvents} from './events';
import Client from './Client';

import './Mobile.css';

class Mobile extends React.PureComponent {
/*
  static propTypes = {
    name: PropTypes.string.isRequired,
    clients:PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fio: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };
*/
  state = {
    name: this.props.name,
    clients: this.props.clients,
    showClients: 'all',
    workMode: 1, //1-отображение, 2 - редактирование, 3 - удаление
    editingClient: null
  };

  componentDidMount = () => {
    delEditEvents.addListener('deleteClient',this.deleteClient);
    delEditEvents.addListener('editClient',this.editClient);
  };

  deleteClient = (clientID) => {
    if(this.state.workMode === 2 || this.state.workMode === 3) {
      return;
    }
    let newClients = this.state.clients.filter(client => 
      client.id !== clientID  
    );
    this.setState({clients: newClients});
  }

  componentWillUnmount = () => {
    delEditEvents.removeListener('deleteClient',this.deleteClient);
    delEditEvents.removeListener('editClient',this.editClient);
  };

  setName1 = () => { 
    this.setState({name:'МТС'});
  };

  setName2 = () => {
    this.setState({name:'Velcom'});
  };
  
  showClientsAll = () => {
    this.setState({showClients: 'all'});
  }

  showClientsActive = () => {
    this.setState({showClients: 'active'});
  }

  showClientsBlocked = () => {
    this.setState({showClients: 'blocked'});
  }

  addClient = () => {
    let newId;
    outer: for(var i = 100;;i++){
      for(var j = 0; j < this.state.clients.length; j++){
        if(i == this.state.clients[j].id){
          continue outer;
        }
        if(j == this.state.clients.length-1){
          newId = i;
          break outer;
        } 
      }
    }
    
    let addingClient = {
      surname: '',
      name: '',
      secName: '',
      balance: '',
      id: newId
    };
    this.setState({workMode: 3,editingClient: addingClient})
  }

  editClient = (client) => {
    if(this.state.workMode === 3 || this.state.workMode === 2) {
      return;
    }
    this.setState({workMode: 2, editingClient: client});
  }
   cancelEditAdd = () => {
     this.setState({workMode: 1, editingClient: null})
   }

   changeClient = (EO) => {
     let changedClient = {...this.state.editingClient};
     if(EO.target.getAttribute('data') === 'surname'){
       changedClient.surname = EO.target.value;
     }
     else if(EO.target.getAttribute('data') === 'name'){
      changedClient.name = EO.target.value;
    }
    else if(EO.target.getAttribute('data') === 'secName'){
      changedClient.secName = EO.target.value;
    }
    else if(EO.target.getAttribute('data') === 'balance'){
      changedClient.balance = EO.target.value;
    }
     this.setState({editingClient: changedClient})
   }
  
  saveChanges = () => {
    let updatedClients = [];
    if(this.state.workMode === 2) {
      updatedClients = this.state.clients.map( client => {
        if(client.id !== this.state.editingClient.id){
          return client;
        }
        else {
          return this.state.editingClient;
        }
      }

      )
    }
    else if(this.state.workMode === 3){
      updatedClients = [...this.state.clients];
      updatedClients.push(this.state.editingClient);
    } 
    this.setState({workMode: 1, editingClient: null, clients: updatedClients})
  }

  render() {

    console.log("Mobile render");

    var clientsCode = [];
    
     this.state.clients.forEach( client => {
        if(this.state.showClients == 'all'){
          clientsCode.push(<Client key={client.id} info={client}/>);
        }
        else if (this.state.showClients == 'active') {
          if(client.balance > 0){
            clientsCode.push(<Client key={client.id} info={client}/>);
          }
        }
        else if(this.state.showClients == 'blocked'){
          if(client.balance <= 0) {
          clientsCode.push(<Client key={client.id} info={client}/>);
          }
        }
     }
    );

    return (
    <div>  
      <div> 
        <button onClick={this.setName1}>МТС</button><button onClick={this.setName2}>Velcom</button>
      </div>
    <hr></hr>  
    <div>Компания: {this.state.name}</div>
    <hr></hr>
    <button onClick={this.showClientsAll}>ВСЕ</button>
    <button onClick={this.showClientsActive}>Активные</button>
    <button onClick={this.showClientsBlocked}>Заблокированные</button>
    <hr></hr>
    <table>
      <thead>
        <tr>
        <th>Фамилия</th><th>Имя</th><th>Отчество</th><th>Баланс</th><th>Статус</th><th>Редактировать</th><th>Удалить</th>
        </tr>
      
      </thead>
      <tbody>
        {clientsCode}
 
      </tbody>
    </table>
    <hr></hr>
    {(this.state.workMode === 1 && <button onClick={this.addClient}>Добавить клиента</button>) ||
    ((this.state.workMode === 2 || this.state.workMode === 3) && 
      <div>
        <h3>{(this.state.workMode === 2)?"Редактирование клиента: "+this.state.editingClient.id:"Добавление нового клиента"}</h3>
        <div>Фамилия: <input value={this.state.editingClient.surname}  data='surname' onChange={this.changeClient}></input></div>
        <div>Имя: <input value={this.state.editingClient.name} data='name' onChange={this.changeClient}></input></div>
        <div>Отчество: <input value={this.state.editingClient.secName} data='secName' onChange={this.changeClient}></input></div>
        <div>Баланс: <input value={this.state.editingClient.balance} data='balance' onChange={this.changeClient}></input></div>
        <button onClick={this.saveChanges}>Сохранить</button>
        <button onClick={this.cancelEditAdd}>Отмена</button>
       </div>    
    ) 
    }
    </div>  
    )
    ;

  }

}

export default Mobile;
