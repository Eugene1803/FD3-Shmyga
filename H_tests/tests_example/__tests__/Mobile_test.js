"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import Client from '../components/Client';
import Mobile from '../components/Mobile';

test('работа Mobile', () => {
  let clientsArr=[ 
    {id:101, surname:"Иванов",name:'Иван',secName: 'Иванович', balance:200}, 
    {id:105, surname:"Петров",name:'Петр',secName: 'Петрович', balance:250}, 
    {id:110, surname:"Сидоров",name:'Сидор',secName: 'Сидорович', balance:180},
    {id:120, surname:"Суворов",name:'Герасим',secName: 'Филиппович', balance:-220},
  ];
  // создаём тестовую версию компонента
  const component = renderer.create(
    <Mobile clients={clientsArr}/>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // найдём в вёрстке компонента саму кнопку
  const buttonElemAll = component.root.find( el => el.type=='button' && el.props.children == 'ВСЕ'); 
  // и "нажмём" на неё
  buttonElemAll.props.onClick();

  // получаем уже изменённый снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // "нажмём" кнопку ещё раз
  const buttonElemActive = component.root.find( el => el.type=='button' && el.props.children == 'Активные'); 
  buttonElemActive.props.onClick();
  
  // и получаем окончательный снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  
  const buttonElemBlocked = component.root.find( el => el.type=='button' && el.props.children == 'Заблокированные'); 
  buttonElemBlocked.props.onClick();
  
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  const buttonElemAdd = component.root.find( el => el.type=='button' && el.props.children == 'Добавить клиента'); 
  buttonElemAdd.props.onClick();
  
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  let buttonElemSave = component.root.find( el => el.type=='button' && el.props.children == 'Сохранить'); 
  buttonElemSave.props.onClick();
  
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  const client = component.root.find('Client');
  console.log(client);
 
  const buttonElemDelete = component.root.find( el => el.type=='button' && el.props.children == 'Удалить'); 
  expect(buttonElemDelete).toBeUndefined()
  /*
  // можно эмулировать события, передавая в качестве объекта события то что нам нужно:
  wrapper.find('select').simulate('change', {
    target: { value: "hello" },
  });
  */

});
