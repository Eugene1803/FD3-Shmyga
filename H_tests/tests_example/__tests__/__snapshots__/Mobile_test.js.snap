// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа Mobile 1`] = `
<div>
  <hr />
  <button
    onClick={[Function]}
  >
    ВСЕ
  </button>
  <button
    onClick={[Function]}
  >
    Активные
  </button>
  <button
    onClick={[Function]}
  >
    Заблокированные
  </button>
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Петр
        </td>
        <td>
          Петрович
        </td>
        <td>
          250
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          180
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Суворов
        </td>
        <td>
          Герасим
        </td>
        <td>
          Филиппович
        </td>
        <td>
          -220
        </td>
        <td
          className="ClientBlocked"
        >
          Заблокирован
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <button
    onClick={[Function]}
  >
    Добавить клиента
  </button>
</div>
`;

exports[`работа Mobile 2`] = `
<div>
  <hr />
  <button
    onClick={[Function]}
  >
    ВСЕ
  </button>
  <button
    onClick={[Function]}
  >
    Активные
  </button>
  <button
    onClick={[Function]}
  >
    Заблокированные
  </button>
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Петр
        </td>
        <td>
          Петрович
        </td>
        <td>
          250
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          180
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Суворов
        </td>
        <td>
          Герасим
        </td>
        <td>
          Филиппович
        </td>
        <td>
          -220
        </td>
        <td
          className="ClientBlocked"
        >
          Заблокирован
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <button
    onClick={[Function]}
  >
    Добавить клиента
  </button>
</div>
`;

exports[`работа Mobile 3`] = `
<div>
  <hr />
  <button
    onClick={[Function]}
  >
    ВСЕ
  </button>
  <button
    onClick={[Function]}
  >
    Активные
  </button>
  <button
    onClick={[Function]}
  >
    Заблокированные
  </button>
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Петр
        </td>
        <td>
          Петрович
        </td>
        <td>
          250
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          180
        </td>
        <td
          className="ClientActive"
        >
          Активен
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <button
    onClick={[Function]}
  >
    Добавить клиента
  </button>
</div>
`;

exports[`работа Mobile 4`] = `
<div>
  <hr />
  <button
    onClick={[Function]}
  >
    ВСЕ
  </button>
  <button
    onClick={[Function]}
  >
    Активные
  </button>
  <button
    onClick={[Function]}
  >
    Заблокированные
  </button>
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Суворов
        </td>
        <td>
          Герасим
        </td>
        <td>
          Филиппович
        </td>
        <td>
          -220
        </td>
        <td
          className="ClientBlocked"
        >
          Заблокирован
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <button
    onClick={[Function]}
  >
    Добавить клиента
  </button>
</div>
`;

exports[`работа Mobile 5`] = `
<div>
  <hr />
  <button
    onClick={[Function]}
  >
    ВСЕ
  </button>
  <button
    onClick={[Function]}
  >
    Активные
  </button>
  <button
    onClick={[Function]}
  >
    Заблокированные
  </button>
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Суворов
        </td>
        <td>
          Герасим
        </td>
        <td>
          Филиппович
        </td>
        <td>
          -220
        </td>
        <td
          className="ClientBlocked"
        >
          Заблокирован
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <div>
    <h3>
      Добавление нового клиента
    </h3>
    <div>
      Фамилия: 
      <input
        data="surname"
        onChange={[Function]}
        value=""
      />
    </div>
    <div>
      Имя: 
      <input
        data="name"
        onChange={[Function]}
        value=""
      />
    </div>
    <div>
      Отчество: 
      <input
        data="secName"
        onChange={[Function]}
        value=""
      />
    </div>
    <div>
      Баланс: 
      <input
        data="balance"
        onChange={[Function]}
        value=""
      />
    </div>
    <button
      onClick={[Function]}
    >
      Сохранить
    </button>
    <button
      onClick={[Function]}
    >
      Отмена
    </button>
  </div>
</div>
`;

exports[`работа Mobile 6`] = `
<div>
  <hr />
  <button
    onClick={[Function]}
  >
    ВСЕ
  </button>
  <button
    onClick={[Function]}
  >
    Активные
  </button>
  <button
    onClick={[Function]}
  >
    Заблокированные
  </button>
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Суворов
        </td>
        <td>
          Герасим
        </td>
        <td>
          Филиппович
        </td>
        <td>
          -220
        </td>
        <td
          className="ClientBlocked"
        >
          Заблокирован
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          
        </td>
        <td>
          
        </td>
        <td>
          
        </td>
        <td>
          
        </td>
        <td
          className="ClientBlocked"
        >
          Заблокирован
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Редактировть
          </button>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <button
    onClick={[Function]}
  >
    Добавить клиента
  </button>
</div>
`;
